[technotes] router authentication (#129)
apollographql/docs: src/content/technotes/TN0004-router-authentication.mdx

---
title: TN0004 - Router Authentication
description: Options for authenticating requests when using Apollo Router
tags: [server, router, auth]
---

This note describes options for authentication when using Apollo Router.

## Authenticate in subgraphs

The simplest architecture is to delegate authentication to subgraph services.

```mermaid
sequenceDiagram
  Client->>Router: Request with token
  Router->>Subgraph: Request with token
  note over Subgraph: Authenticate request with token
```

> The YAML configuration for forwarding the `Authorization: Bearer jsonwebtoken` header:

Reviewer: Because auth can cover either an opaque token or JWTs, abstract a little. 

```suggestion
The YAML configuration for forwarding the `Authorization` header:
```

```yaml
headers:
  all:
    - propagate:
        named: authorization
```

**Pros**

- Requires very little Router configuration.

> - Enforces authentication at ingress instead of in each subgraph.

Author: whoops, copy-pasted into the wrong section!

- Can leverage existing authentication code, which may be closely tied to authorization logic close to data sources.

**Cons**

- Each subgraph re-authenticates the actor with each subgraph fetch.
- If subgraphs are written in different languages, maintaining authentication code for each is not ideal.

> ## Envoy ext_autz

Reviewer: Question: Is it worth abstracting this a bit (or at least note) that you could do auth before hitting GQL? Envoy does this, but someone could use another tech stack if they're more familiar with it.

[Envoy](https://www.envoyproxy.io/) is a powerful proxy with [built-in support for authenticating and authorizing requests with an external service](https://www.envoyproxy.io/docs/envoy/v1.22.0/intro/arch_overview/security/ext_authz_filter). See this [demo project](https://github.com/apollosolutions/router-ext-authz) for an example that uses [Open Policy Agent](https://www.openpolicyagent.org/) as the Policy Decision Point.

```mermaid
sequenceDiagram
  Client->>Envoy: Request with token
  Envoy->>OPA: Request with token

>   OPA->>JWKS: /.well-known/jwks

Reviewer: It might be worth a note stating that this example is for JWTs either at the top level, or within this section; for those not using JWTs, JWKS might not be an acronym/tech they're aware of or know of.

  JWKS->>OPA: { "keys": [...] }
  note over OPA: Verify token, extract claims
  OPA->>Envoy: 200 OK + new headers
  Envoy->>Router: Request with new headers
  Router->>Subgraph: Request with new headers
```

**Pros**

- Requires very little Router configuration.
- The authorization service can execute arbitrary logic. For example, it can extract claims from the JWT, converting them to headers that are forwarded to subgraphs.
- Envoy (as well as Open Policy Agent) typically run as sidecars within the same pod as the router, so requests between services are very fast.

**Cons**

- Does not support both authenticated and unauthenticated APIs (for example, a `login` mutation).
- Requires Envoy and a control plane for configuring the `ext_authz` filter.

## Router extension

The Router can authenticate requests in a custom plugin. See this [example in the router repository](https://github.com/apollographql/router/tree/main/examples/jwt-auth) for an demonstration of JWT authentication.

```mermaid
sequenceDiagram
  Client->>Router: Request with token
  Router->>JWKS: /.well-known/jwks

>   JWKS->>Router: { "keys": [...] }

Reviewer: The example above is actually not JWKS signed-JWTs, but instead symmetrically-signed JWTs where there's a single, static key used.

Author: can you elaborate on the difference? is there a clarification i should make, or a change i should make to the demo?

Author: ohhhh i see what you're saying. i couldn't tell the context from the code above your comment.

Reviewer: Ah! I could've selected more lines there- sorry about that!

Author: great catch, simplified

  note over Router: Verify token
  note over Router: Add headers to context
  Router->>Subgraph: Request with headers from context
```

**Pros**

- Enforces authentication at ingress instead of in each subgraph.
- Doesn't require additional services, sidecars, or control planes.
- The plugin can execute arbitrary logic. For example, it can extract claims from the JWT, attaching them to the request context so they can be forwarded to subgraphs.

**Cons**

> - Requires extensive customization of the Router.

Reviewer: ```suggestion
- Requires extensive customization of the Router in Rust.
```

Reviewer: only clarifying as it can't be done in Rhai just yet (afaik).

---
title: TN0004 - Router Authentication

> description: Options for authenticating requests when using Apollo Router

Reviewer 2: (Just seems like we can fold most of the language in the description up into the title)

tags: [server, router, auth]
---

This note describes options for authentication when using Apollo Router.

## Authenticate in subgraphs

The simplest architecture is to delegate authentication to subgraph services.

```mermaid
sequenceDiagram
  Client->>Router: Request with token
  Router->>Subgraph: Request with token
  note over Subgraph: Authenticate request with token
```

> The YAML configuration for forwarding the `Authorization` header:

Reviewer 2: ```suggestion
To pass `Authentication` headers from client requests to your subgraphs, add the following to your router's YAML configuration file:
```

```yaml
headers:
  all:
    - propagate:
        named: authorization
```

**Pros**

> - Requires very little Router configuration.

Reviewer 2: ```suggestion
- Requires minimal changes to your router configuration.
```

> - Can leverage existing authentication code, which may be closely tied to authorization logic close to data sources.

Reviewer 2: ```suggestion
- Can take advantage of existing authentication code in subgraphs, which is often tied to authorization logic for data sources.
```

**Cons**

> - Each subgraph re-authenticates the actor with each subgraph fetch.

Reviewer 2: ```suggestion
- Each subgraph that contributes to resolving a request needs to authenticate that request.
```

> - If subgraphs are written in different languages, maintaining authentication code for each is not ideal.

Reviewer 2: ```suggestion
- If subgraphs are written in different languages, maintaining consistent authentication code for each is complex.
```

## Ingress filter

> Many ingresses or generic API gateways have authentication features.

Reviewer 2: ```suggestion
Many ingresses and generic API gateways provide authentication features.
```

[Envoy](https://www.envoyproxy.io/) is a powerful proxy with [built-in support for authenticating and authorizing requests with an external service](https://www.envoyproxy.io/docs/envoy/v1.22.0/intro/arch_overview/security/ext_authz_filter). See this [demo project](https://github.com/apollosolutions/router-ext-authz) for an example that uses [Open Policy Agent](https://www.openpolicyagent.org/) as the Policy Decision Point.

```mermaid
sequenceDiagram
  Client->>Envoy: Request with token
  Envoy->>OPA: Request with token
  OPA->>Key server: Get keys
  Key server->>OPA: Public keys
  note over OPA: Verify token
  note over OPA: Add claims as headers
  OPA->>Envoy: 200 OK + new headers
  Envoy->>Router: Request + new headers
  Router->>Subgraph: Request + new headers
```

**Pros**

> - Requires very little Router configuration.

Reviewer 2: ```suggestion
- Requires minimal changes to your router configuration.
```

> - Enforces authentication at ingress instead of in each subgraph.

Reviewer 2: ```suggestion
- Enforces authentication at a single point, reducing complexity relative to authenticating in each subgraph.
```

- The authorization service can execute arbitrary logic. For example, it can extract claims from a JSON Web Token (JWT), converting them to headers that are forwarded to subgraphs.
- Envoy (as well as Open Policy Agent) typically run as sidecars within the same pod as the router, so requests between services are very fast.

**Cons**

- Does not support both authenticated and unauthenticated APIs (for example, a `login` mutation).
- Requires Envoy and a control plane for configuring the `ext_authz` filter.

## Router extension

> The Router can authenticate requests in a custom plugin. See this [example in the router repository](https://github.com/apollographql/router/tree/main/examples/jwt-auth) for an demonstration of JWT authentication.

Reviewer 2: ```suggestion
The Apollo Router can authenticate requests in a custom plugin. For a demonstration of JWT authentication, see [this example in the router repository](https://github.com/apollographql/router/tree/main/examples/jwt-auth).
```

```mermaid
sequenceDiagram
  Client->>Router: Request with token
  note over Router: Verify token with HMAC
  note over Router: Add headers to context
  Router->>Subgraph: Request with headers from context
```

**Pros**

> - Enforces authentication at ingress instead of in each subgraph.

Reviewer 2: ```suggestion
- Enforces authentication at a single point, reducing complexity relative to authenticating in each subgraph.
```

- Doesn't require additional services, sidecars, or control planes.
- The plugin can execute arbitrary logic. For example, it can extract claims from a JWT, attaching them to the request context so they can be forwarded to subgraphs.

**Cons**

> - Requires extensive customization of the Router in Rust.

Reviewer 2: ```suggestion
- Requires extensive customization of the router in Rust.
```
